src/python/pants/backend/python/rules/resolve_requirements.py

```diff
@@ -0,0 +1,90 @@
+# coding=utf-8
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from future.utils import text_type
+
+from pants.backend.python.subsystems.python_native_code import PexBuildEnvironment, PythonNativeCode
+from pants.backend.python.subsystems.python_setup import PythonSetup
+from pants.engine.fs import Digest, Snapshot, UrlToFetch
+from pants.engine.isolated_process import ExecuteProcessRequest, ExecuteProcessResult
+from pants.engine.rules import optionable_rule, rule
+from pants.engine.selectors import Get
+from pants.util.objects import datatype, hashable_string_list, string_optional, string_type
+from pants.util.strutil import create_path_env_var
+
+
+class ResolveRequirementsRequest(datatype([
+  ('requirements', hashable_string_list),
+  ('output_filename', string_type),
+  ('entry_point', string_optional),
```

Comment:
If the entrypoint is optional, should it be the last argument? But also, it doesn't seem like a requirements pex should ever need an entry_point (unless there were no sources, I suppose? but probably not worth special casing).

Reply:
We need the entry point for `python_test_runner.py`, because it sets the entry point `pytest:main`.

That's fine with moving it to the last member of this datatype, although worth noting that `interpreter_constraints` is also ~optional in the sense that you can give an empty tuple if there are no constraints. I'd rather instead be able to make it an `Option[List[str]`, meaning it can be `None`, but don't know how to do that in the current type system.